### ISSUE TITLE
Make a function to extract anisotropy versus radial coordinate

### ISSUE THREAD

**Issue by DanHickstein** [PyAbel/PyAbel#176]
I got an e-mail from a PyAbel user a few days ago: he was accustomed to using the BASEX software and he wanted to calculate the anisotropy parameter as a function of the radial coordinate. Currently, our `abel.tools.vmi.anisotropy_parameters()` function only calculates the anisotropy parameter for one specific curve. While it is possible to pass several radial ranges to `abel.tools.vmi.radial_integration()`, it's not possible to get the "anisotropy-versus-radius curve" without a little bit of coding. 

So, the question is, should this just be a function in PyAbel? It's only a few lines of code, but it might save interested users some time.

I provided the user following code that achieves this, and it could be wrapped up into a function pretty easily:

```python
# We start with AIM as the transformed image.
# first we set up the ranges for radial integration

r = np.arange(0,300,2) # the step here should be changed to give good signal-to-noise
r_range = zip(r[:-1], r[1:])

# map to intensity vs theta for each radial range
intensities, theta = abel.tools.vmi.radial_integration(AIM, radial_ranges=r_range)


# The following just loops over the intensity values and calculated the anisotropy
# There is probably a way more Pythonic way to do this. Probably with the "map" function...
anisotropies = np.zeros(np.shape(intensities))
errors       = np.zeros(np.shape(intensities))

print("radial-range      anisotropy parameter (beta)")
for i, (rr, intensity) in enumerate( list(zip(r_range, intensities)) ):

    (beta, error_beta), (amplitude, error_amplitude) = abel.tools.vmi.anisotropy_parameter(theta, intensity)
    
    anisotropies[i] = beta
    errors[i] = error_beta
```

**Comment by stggh** [comment]
@DanHickstein this is very good, and your code `r_range = zip(r[:-1], r[1:])` is an education.

I was lead astray in the past by the notion that the (original) `basex.exe` returned anisotropy parameter was [nonsense](https://github.com/PyAbel/PyAbel/issues/167#issuecomment-214210305). However, sensible values are obtained if the Newton spheres associated with low intensities are eliminated (the `threshold` parameter in `linbasex`). The fact that the anisotropy parameter is directly available from the `linbasex` and `basex` calculations is a neat feature. If possible, in PyAbel,  a direct evaluation of the anisotropy parameter should also be made available for `basex`. Over to you @DanHickstein or @DhrubajyotiDas?  While revisiting the `basex` code please consider making available the full image transform function, as it once was. This may then be called by the `abel.Transform()` instance in the case where `symmetry_axis=None`. See the the code  for `linbasex` to see how this is done.:
```python
    def _abel_transform_image_full(self, **transform_options):

        abel_transform = {
            # "basex": basex.basex_transform,
            "linbasex": linbasex.linbasex_transform_full
        }

```

**Comment by DanHickstein** [comment]
Yeah, if there is no signal, then the anisotropy is nonsense. By default, I think that a function should pass back all of the anisotropy parameters, but we could perhaps also add a feature to return a masked array that is only valid at intensity values over some threshold? This seems complicated - perhaps better to just let the user set their own intensity threshold via a simple example in the docstring of this proposed function. 

I think that I have some old Matlab code where the anisotropies are calculated analytically from the basis set coefficients for basex (it's in the PyAbel/Legacy repo). As I recall, this was painfully slow, and I'm not sure if you actually gain anything. The guassian functions used as the basis set are 1 pixel in size, so there should likely be little to gain from using them directly rather than the current integration method.

I maybe need to think a little bit more about the "returning the full image" issue that you mention. Fundamentally, doesn't the linbasex method assume at least left-right symmetry? The basex code assumed left/right, so returning the full image would really correspond to a symmetry_axis=(0) scenario. What is the motivation for returning the full image?

**Comment by stggh** [comment]
The fundamental image "block" processing size for our transform methods differ:

- (right-side) 1/2-image:

    `hansenlaw`, `two_point`, `three_point`, `onion_peeling`, `direct`(?), `onion_bordas`(?) 

- full-image: 

    `linbasex` (Lin_Basex), and `basex` 

`basex` and `linbasex` conform to the 1/2-image paradigm by forming a full image from two identical 1/2-images, that is then transformed, with 1/2-image returned. For a consistent interface this is fine.  It would be good to have them process only 1/2-image, but as @gerber-t states ["To what avail?"](https://github.com/PyAbel/PyAbel/issues/167#issuecomment-216201576).

For `abel.Transform()` why just not call the full-image method if suitable (when no image symmetry is selected) and available?

The issue is there is a great deal of "messing around" for the quadrant image processing, particularly, in reassembling  the `Beta` arrays of `linbasex`. There is also the loss of transform speed, in that each image may be transformed 4x (each quadrant).

My argument is that `abel.Transform()` should choose the most appropriate and efficient transform mechanism. For `hansenlaw` et al. this may also include 1/2-image processing (rather than 1/4).

**Comment by DanHickstein** [comment]
@stggh, I think I see what you are saying, that it seems a little silly to split images apart and re-assemble them when some methods are returning "full images".

Indeed, we have given the `abel.Transform` class the unusual ability: it can return inverse Abel transforms that do not exhibit cylindrical symmetry! (The left side is different from the right side.)

However, if `symmetry=None`, then splitting into quadrants is the only way! All of the transform methods should assume symmetry around the vertical axis.

I think that it's worth separating the symmetries that are fundamental to the transform approach from those that are idiosyncrasies of the exact implementation. Fundamentally, all of the Abel transform methods should be assuming cylindrical symmetry, since this is the basic assumption of the Abel transform itself. On this level, there should really be no such thing as an Abel transform method that accepts a "full image", consisting of an asymmetric left and right side. Basex is basically a "right side" method, since it ends up averaging the left and right sides. It generates a "full image" with symmetry about the vertical axis. It should be possible to modify the basex code to work on just one side of the image, and in the long run, I think that this should be done.

The same applies to `linbasex`, it assumes symmetry around the vertical axis, so the result should be equivalent to simply averaging the two sides before processing. Of course, the actual implementation that we are using operates on a full image, but fundamentally it shouldn't need to. 

I agree that the breaking up into quadrants and symmetrizing images is a little bit confusing, and adds to the codebase. However, what we have now works well, and I am not sure if adding additional complexity via different processing pathways is the best way to simplify things.

Linbasex is an interesting case, because if offers the possibility to generate information beyond the 2D transformed image (the betas). It really only makes sense to look at the betas in the case where the linbasex algorithm is run only once. I think that the way that you have currently implemented this in the abel.Transform method is okay, but it adds a lot of code. One option would be to keep abel.Transform simple: it just applies the methods in quadrants and returns the 2D images. If the users want to do more, then they can refer to the individual transform functions. 

However, I think that there will be future methods that also use anisotropies, and perhaps it's not a bad thing to have abel.Transform be a centralized function, even if it makes it more complicated.

Anyway, in the end, I think that the way we are currently dealing with things works fairly well. I don't see a need to modify Basex to handle full images unless this opens additional capabilities, as it does for linbasex.

**Comment by stggh** [comment]
I am happy with keeping the quadrant and symmetry processing. However, it is trival for `abel.Transform()` to call the most appropriate function if available. `basex` once processed the full image. I would prefer the function put back, in *addition* to what is there.  I think that the change would be relatively simple, just providing the function name (wrapper), @DhrubajyotiDas ?

In `abel.Transform` the extra code for full image transform is small (most lines associated with trace printing):
```python
    def _abel_transform_image(self, **transform_options):
        if self.method == "linbasex" and self._symmetry_axis is not None:
            self._abel_transform_image_full(**transform_options)
        else:
            self._abel_transform_image_by_quadrant(**transform_options)


    def _abel_transform_image_full(self, **transform_options):

        abel_transform = {
            # "basex": basex.basex_transform,
            "linbasex": linbasex.linbasex_transform_full
        }
        t0 = time.time()

        self._verboseprint('Calculating {0} Abel transform using {1} method -'
                          .format(self.direction, self.method),
                          '\n    image size: {:d}x{:d}'.format(*self.IM.shape))

        self.transform, self.radial, self.Beta, self.projection =\
                                abel_transform[self.method](self.IM, **transform_options)

        self._verboseprint("{:.2f} seconds".format(time.time()-t0))

 ```

Tthe core functionality of `basex` is the full-image transform. The function could be returned, at least as a direct call function. In the end I am not fussed, and @DanHickstein is usually right ;-).

**Comment by DanHickstein** [comment]
> The core functionality of basex is the full-image transform

This is true, but only because we were too lazy to change it :). In the future, I hope that it will be changed to just a single side (it should speed it up a little bit), so I'd rather not implement any code whereby BASEX completes the "both sides" transform, unless there is a compelling reason to do so.

**Comment by stggh** [comment]
I only want the (additional) function statement (wrapper) returned as it once was.  I am pretty sure `basex` processes the whole image, just the function was renamed.  @DhrubajyotiDas any comment?
Not a big deal.

**Comment by DanHickstein** [comment]
Yeah, the current implementation does process the full image, but I would like to someday remedy that so that we can generate basis sets that are 1/2 the size they are now.

**Comment by stggh** [comment]
In the `linbasex` case @gerber-t has said there is little point in processing 1/2 images ["To no avail"](https://github.com/PyAbel/PyAbel/issues/167#issuecomment-216201576), but perhaps, he meant this in the context of "why bother with the extra work".  1/2-sized basis sets would definitely be a good thing.

**Comment by gerber-t** [comment]
@stggh 
Exactly that was I ment. 
In addition, linbasex lives from making projections along some axis at some angles (along the x -axis alone, or along the x- and y- axis are only the most simple cases. You may want to explore more axes at more esoteric angles to take profit of averaging). For making these projections you need the full image. After having the projections you may proceed with only half length arrays if (IF!) there are only even symmetries (no uneven Legendre polynomials). If you a priori know that there are uneven symmetries, then you need the full image and full projections anyway.

**Comment by DhrubajyotiDas** [comment]
@stggh Apologies for the late response.

>I am pretty sure basex processes the whole image, just the function was renamed.

Yes, and as @DanHickstein mentioned, technically the `basex`algorithm can be recast to do just one-half natively (although that's still on my to-do list and will be taken care of someday...) 

>. I think that the change would be relatively simple, just providing the function name (wrapper),

The relevant code that would need to be changed/circumvented is [here](https://github.com/PyAbel/PyAbel/blob/master/abel/basex.py#L119). There, the half image is being mirrored to a full image and then processed. For what @stggh wants to accomplish, you'd need to make `full_image = data` in some pythonic way.

The simplest (if not the most elegant) way I can think of right now either by duplicating the entire `basex_transform` function without [Line 119](https://github.com/PyAbel/PyAbel/blob/master/abel/basex.py#L119).

If you're currently working on a PR, feel free to make this change along with the rest.

**Comment by DanHickstein** [comment]
> technically the basexalgorithm can be recast to do just one-half natively (although that's still on my to-do list and will be taken care of someday...)

So, I view to two-sided version of BASEX to basically be depreciated.

What can you accomplish with the left+right side basex transform that could not be accomplished by completing the transforms separately?

**Comment by DhrubajyotiDas** [comment]
Nothing specific comes to mind. The left+right side is useful in that it wraps the averaging-left-and-right-side into one step, but if we are splitting the image into two halves and passing each to `basex` right now, we are not getting that benefit anyway.

**Comment by stggh** [comment]
I think that I have created a mountain out of a mole-hill, and stolen the topic of this issue, sorry. At least there is now one less "to do" job for @DhrubajyotiDas  ;-).

The answer is trivial, remove line 119. For the case of 1/2-image transforms, apply 119  to the 1/2 image prior to calling `basex_transform()`

However, as this not popular, I have no issue in leaving `basex` as it is. No worries.

**Comment by stggh** [comment]
Discussion  [moved from PM] concerning the accuracy of anisotropy parameters extracted via integration over selected spectral features (those with good intensity) vs every radial radial coordinate, using @DanHickstein's clever technique mentioned in the [preamble](https://github.com/PyAbel/PyAbel/issues/176#issue-153615099):

>On Wed, May 11, 2016 at 7:48 PM, @DanHickstein  wrote:
Yes, the anisotropy seems like a difficult quality to tease out, because basically you need to end up fitting a function to noisy data, which often doesn't work so well. I wonder if, from the lineout of intensity vs angle, if one could just calculate statistics such as the standard deviation and kurtosis.

>On Friday, May 13, 2016, @DanHickstein:
we could probably write a function for an example image that consists of randomly samples gaussian functions with radii of a few pixels (detector hits) dispersed according to a larger function (gausssian)

PyAbel has the synthetic images `abel.tools.analytical.sample_image(name='dribinski')` and `name=Ominus`, together with the `linbasex`  `examples/data/VMI-art1.txt.bz2`, which may be enough. I will also compare the results of the two techniques with the real experiment data `examples/data/O2-ANU1024.txt.bz2`.

In relation to you code, I believe that it may easily be incorporated into `abel.tools.vmi.radial_integration()` with the default replaced with your `zip`:
```python
    if radial_ranges is None:
        radial_ranges = [(0, r[-1]), ]
```
to something like:
```python
    if radial_ranges is None:
        radial_ranges = 1
    if isinstance(radial_range, int):
        radial_ranges = zip(r[:-radial_ranges], r[radial_ranges:])
```
(not tested!)

**Comment by DanHickstein** [comment]
I'll open a separate issue about the sample images - basically we currently have no way of adding noise to our sample images, and this would be very nice for testing purposes.

The idea to make the default behavior of radial integration to integrate many bins seems like a good one. Still, we would want a separate function for the anisotropy parameter as a function of `r`, right? Or, are you suggesting that we simple add this ability to the existing `abel.tools.vmi.anisotropy_parameter` function?

**Comment by stggh** [comment]
Moved to PR #178 

@DanHickstein note you have an outstanding "open a separate issue about the sample images - basically we currently have no way of adding noise"
